File: go/internal/anagram/anagram.go
package anagram

import (
	"strconv"
	"strings"
)

// Exec groups anagrams from an array of strings.
// An anagram is a word formed by rearranging the letters of another word. E.g.: "earth" - "heart"
//
// Example:
// Input: words = ["earth", "arrest", "program", "traces", "heart"]
// Output: [["earth", "heart"], ["arrest", "traces"], ["program"]]
//
// Returns an array of grouped anagrams. Any order is valid.
func Exec(words []string) [][]string {
	if len(words) == 0 {
		return [][]string{}
	}

	groups := buildGroup(words)

	return groups
}

func buildGroup(words []string) [][]string {
	groups := make(map[string][]string)

	for _, word := range words {
		hash := hashWord(word)

		groups[hash] = append(groups[hash], word)
	}

	result := make([][]string, 0, len(groups))

	for _, group := range groups {
		result = append(result, group)
	}

	return result
}

func hashWord(word string) string {
	hash := [26]int{}

	for _, char := range word {
		index := int(char - 'a')

		hash[index]++
	}

	var builder strings.Builder

	for _, num := range hash {
		builder.WriteString(strconv.Itoa(num))
	}

	return builder.String()
}

File: tsc/src/dsa/anagram/anagram.ts
/**
 * Groups anagrams from an array of strings.
 * An anagram is a word formed by rearranging the letters of another word. E.g.: "earth" - "heart"
 *
 * @param words - Array of strings to group
 * @returns Array of grouped anagrams. Any order is valid.
 *
 * @example
 * Input: words = ["earth", "arrest", "program", "traces", "heart"]
 * Output: [["earth", "heart"], ["carest", "traces"], ["program"]]
 */
export const groupAnagrams = (words: string[]): string[][] => {
  if (!words || !Array.isArray(words) || words.length === 0) {
    return []
  }

  const group = buildGroup(words)

  return group.map((words) => words.sort())
}

const buildGroup = (words: string[]) => {
  const group: Record<string, string[]> = {}

  for (let word of words) {
    const hash = getWordHash(word)

    if (!group[hash]) {
      group[hash] = []
    }

    group[hash].push(word)
  }

  return Object.values(group)
}

const getWordHash = (word: string): string => {
  const charsCount: number[] = new Array(26).fill(0)

  for (let i = 0; i < word.length; i++) {
    const char = word.charCodeAt(i) - 'a'.charCodeAt(0)
    charsCount[char]++
  }

  return charsCount.join('')
}
